sycl/include/sycl/buffer.hpp

```diff
@@ -629,6 +629,8 @@ class buffer : public detail::buffer_plain,
   template <typename Destination>
   detail::EnableIfOutputIteratorT<Destination>
   set_final_data_internal(Destination FinalData) {
+    static_assert(!std::is_const_v<std::remove_reference_t<decltype(*FinalData)>>,
+                  "set_final_data must be called with a non-const iterator!");
```

Comment:
Why this and not the fix in `EnableIfOutputIteratorT`?

Reply:
My reasoning was that by adding it to `EnableIfOutputIteratorT`, when we call set_final_data_internal with a const iterator it simply gets ignored by overload resolution and the compiler diagnostics will most likely be confusing for someone who is not that familiar with template errors. Furthermore, this is an easy mistake to make when writing SYCL code and therefore it would be nice to get a clear error message telling you what you did wrong and thats what a static_assert does. If you're not convinced however, I can add it to `EnableIfOutputIteratorT`.

Reply:
>  it simply gets ignored by overload resolution 

And that's what the spec says should happen, isn't it?

Reply:
Can you provide the part of the spec that says this?
All I can see is that the spec says what to do if the iterator is not const, I cant see anything regarding what to do if it is const.

Reply:
SYCL:
> Destination can be either an output iterator or a std::weak_ptr<T>.

C++: 
> The output_iterator concept is a refinement of [input_or_output_iterator](https://en.cppreference.com/w/cpp/iterator/input_or_output_iterator), adding the requirement that it can be used to write values of type and value category encoded by T (via [indirectly_writable](https://en.cppreference.com/w/cpp/iterator/indirectly_writable)). [equality_comparable](https://en.cppreference.com/w/cpp/concepts/equality_comparable) is not required.

Reply:
I see, thanks for the clarification!
I will make the suggested changes!